snorkel/features/relative_features.py

```diff
@@ -0,0 +1,190 @@
+    for i, arg in enumerate(candidate.get_contexts()):
+        for token in re.split(split_pattern, arg.get_span().lower()):
+            if stopwords is None or token not in stopwords:
+                yield 'SPAN_SPLIT[{0}][{1}]'.format(i, token), 1
+
+
+def get_span_splits_stopwords(stopwords):
+    """Get a span split unary function"""
+    return partial(get_span_splits, stopwords=stopwords)
+
+
+def get_unary_span_feats(sidxs, sentence, stopwords):
+    """Get unary span features from DDLib and TreeDLib"""
+    get_tdl_feats = compile_entity_feature_generator()
+    sent_dict     = get_as_dict(sentence)
+    xmltree       = corenlp_to_xmltree(sent_dict)
+    if len(sidxs) > 0:
+        # Add DDLIB entity features
+        for f in get_ddlib_feats(sent_dict, sidxs):
+            yield 'DDL_' + f, 1
+        # Add TreeDLib entity features
+        for f in get_tdl_feats(xmltree.root, sidxs, stopwords=stopwords):
+            yield 'TDL_' + f, 1
+
+
+def get_binary_span_feats(sidxs, sentence, stopwords):
+    """Get binary (relation) span features from TreeDLib"""
+    get_tdl_feats = compile_relation_feature_generator()
+    xmltree = corenlp_to_xmltree(get_as_dict(sentence))
+    s1_idxs, s2_idxs = sidxs
+    if len(s1_idxs) > 0 and len(s2_idxs) > 0:
+        # Apply TDL features
+        for f in get_tdl_feats(xmltree.root, s1_idxs, s2_idxs,
+                               stopwords=stopwords):
+            yield 'TDL_' + f, 1
+
+
+def get_span_feats(candidate, stopwords=None):
+    """Base function for sentence dependency path features
+    candidate: @Candidate to extract features for
+    stopwords: @set of stopwords to filter out from dependency path
+    """
+    args = candidate.get_contexts()
+    if not isinstance(args[0], Span):
+        raise ValueError("Accepts Span-type arguments, %s-type found.")
+    # Unary candidates
+    if len(args) == 1:
+        sidxs = range(args[0].get_word_start(), args[0].get_word_end() + 1)
+        return get_unary_span_feats(sidxs, candidate.get_parent(), stopwords)
+    # Binary candidates
+    elif len(args) == 2:
+        sidxs = [range(a.get_word_start(), a.get_word_end() + 1) for a in args]
+        return get_binary_span_feats(sidxs, candidate.get_parent(), stopwords)
+    else:
+        raise NotImplementedError("Only handles unary or binary candidates")
+
+
+def get_span_feats_stopwords(stopwords):
+    """Get a span dependency tree unary function"""
+    return partial(get_span_feats, stopwords=stopwords)
+
+
+def get_entity_word_idxs(sentence, et, cid):
+    """Get indices of @sentence tokens with type @et and id @cid"""
+    itr = enumerate(zip(sentence.entity_types, sentence.entity_cids))
+    return [i for i, (t, c) in itr if c == cid and t == et]
+
+
+def get_first_document_span_feats(candidate, stopwords=None):
+    """Base function for sentence dependency path features where the sentence
+       is the first sentence in the parent @Document of @candidate which
```

Comment:
Agree it should be generalized, but is practically useful enough to not be arbitrary. Let's generalize if we hear of a related use case. Can't think of one right now.